cura/MachineActionManager.py

```diff
@@ -0,0 +1,142 @@
+
+            required_actions = container.getMetaDataEntry("required_actions", [])
+            for action in required_actions:
+                self.addRequiredAction(container.getId(), action)
+
+            first_start_actions = container.getMetaDataEntry("first_start_actions", [])
+            for action in first_start_actions:
+                self.addFirstStartAction(container.getId(), action)
+
+    ##  Add a required action to a machine
+    #   Raises an exception when the action is not recognised.
+    def addRequiredAction(self, machine_id, action_key):
+        if action_key in self._machine_actions:
+            if machine_id in self._required_actions:
+                self._required_actions[machine_id] |= {self._machine_actions[action_key]}
+            else:
+                self._required_actions[machine_id] = {self._machine_actions[action_key]}
+        else:
+            raise UnknownMachineAction("Action %s, which is required for %s is not known." % (action_key, machine_id))
+
+    ##  Add a supported action to a machine.
+    def addSupportedAction(self, machine_id, action_key):
+        if action_key in self._machine_actions:
+            if machine_id in self._supported_actions:
+                self._supported_actions[machine_id] |= {self._machine_actions[action_key]}
+            else:
+                self._supported_actions[machine_id] = {self._machine_actions[action_key]}
+        else:
+            Logger.log("w", "Unable to add %s to %s, as the action is not recognised", action_key, machine_id)
+
+    ##  Add an action to the first start list of a machine.
+    def addFirstStartAction(self, machine_id, action_key, index = None):
+        if action_key in self._machine_actions:
+            if machine_id in self._first_start_actions:
+                if index is not None:
+                    self._first_start_actions[machine_id].insert(index, self._machine_actions[action_key])
+                else:
+                    self._first_start_actions[machine_id].append(self._machine_actions[action_key])
+            else:
+                self._first_start_actions[machine_id] = [self._machine_actions[action_key]]
+        else:
+            Logger.log("w", "Unable to add %s to %s, as the action is not recognised", action_key, machine_id)
+
+    ##  Add a (unique) MachineAction
+    #   if the Key of the action is not unique, an exception is raised.
+    def addMachineAction(self, action):
+        if action.getKey() not in self._machine_actions:
+            self._machine_actions[action.getKey()] = action
+        else:
+            raise NotUniqueMachineAction("MachineAction with key %s was already added. Actions must have unique keys.", action.getKey())
+
+    ##  Get all actions supported by given machine
+    #   \param machine The machine you want the supported actions of
+    #   \returns set of supported actions.
+    @pyqtSlot(str, result = "QVariantList")
+    def getSupportedActions(self, machine_id):
```

Comment:
To prevent confusion between stacks and definitions, I think it would be better to name the argument ```definition_id``` instead of ```machine_id```